### PR TITLE
Implement go to definition and document symbols

### PR DESCRIPTION
**Pull request by olafurpg**
"Peek definition"
screen shot 2017-11-05 at 22 27 49

"Go to definition" (right before clicking)
screen shot 2017-11-05 at 22 28 42

"Document symbols"
screen shot 2017-11-05 at 22 30 06

Go to definition still fails on a few cases related to synthetic case class companion object and `B` in `import a.B`, those should be fairly simple to fix and would be good contributions for someone who wants to learn about scalameta/semanticdb.

This PR refactors the Scalafix service so that
- it uses monix.Observable that subscribes to same source as the SymbolIndexer
- it's named `scalafix: Linter` in line with `scalafmt: Formatter`

Monix is pretty awesome, I can imagine see how we can use more of the Observable utilities/overflow-strategies to debounce duplicate dialogues (eg. "Missing .scalafix.conf") or kill stale unhandled requests in favor of latest requests.

### CODE REVIEW

File: language-server/src/main/scala/scala/meta/languageserver/ScalametaEnrichments.scala [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+      else if (isClass) SymbolKind.Class
+      else if (isObject) SymbolKind.Module
+      else if (isTrait) SymbolKind.Interface
+      else if (isPackage || isPackageObject) SymbolKind.Package
+      else if (isType) SymbolKind.Namespace
+      else SymbolKind.Variable // ???
+    }
+  }
+  implicit class XtensionInputLSP(val input: m.Input) extends AnyVal {
+    def contents: String = input.asInstanceOf[m.Input.VirtualFile].value
```
Comment (gabro) [inline language-server/src/main/scala/scala/meta/languageserver/ScalametaEnrichments.scala]: is this `asInstanceOf` safe?

File: language-server/src/main/scala/scala/meta/languageserver/ScalametaEnrichments.scala [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  }
+  implicit class XtensionAbsolutePathLSP(val path: m.AbsolutePath)
+      extends AnyVal {
+    def toLocation(pos: m.Position): l.Location =
+      l.Location(path.toLanguageServerUri, pos.toRange)
+    def toLanguageServerUri: String = "file:" + path.toString()
+  }
+  implicit class XtensionPositionRangeLSP(val pos: m.Position) extends AnyVal {
+    def toRange: l.Range = l.Range(
+      l.Position(line = pos.startLine, character = pos.startColumn),
```
Comment (gabro) [inline language-server/src/main/scala/scala/meta/languageserver/ScalametaEnrichments.scala]: are these lines numbers correct? IIRC there's an off-by-one between LSP and Scala

File: language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala [hunk 3]
```diff
@@ -5,52 +5,87 @@ import java.io.OutputStream
...
+    lspIn: InputStream,
+    lspOut: OutputStream,
+    stdout: PrintStream
+)(implicit s: Scheduler)
+    extends LanguageServer(lspIn, lspOut) {
+  implicit val workspacePath: AbsolutePath = cwd
+  val (semanticdbSubscriber, semanticdbPublisher) =
+    ScalametaLanguageServer.semanticdbStream
+  val buffers: Buffers = Buffers()
+  val symbol: SymbolIndexer =
```
Comment (gabro) [inline language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala]: I would call this `symbolIndexer`. `symbol` is a bit confusing, given the context

File: language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala [hunk 4]
```diff
@@ -62,17 +97,20 @@ class ScalametaLanguageServer(
...
-        path.toString()
-      )
-      List(TextEdit(fullDocumentRange, formattedContent))
+      val config = cwd.resolve(".scalafmt.conf")
+      if (Files.isRegularFile(config.toNIO)) {
+        val formattedContent =
+          scalafmt.format(contents, config.toString(), path.toString())
+        List(TextEdit(fullDocumentRange, formattedContent))
+      } else {
+        connection.showMessage(MessageType.Info, s"Missing $config")
```
Comment (gabro) [inline language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala]: Can't we just fallback to the default config? I believe that's what most plugins do, right?

File: language-server/src/main/scala/scala/meta/languageserver/SymbolIndexer.scala [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+import langserver.core.Connection
+import langserver.messages.MessageType
+
+// NOTE(olafur) it would make a lot of sense to use tries where Symbol is key.
+class SymbolIndexer(
+    val indexer: Observable[Unit],
+    logger: Logger,
+    connection: Connection,
+    buffers: Buffers,
+    documents: JMap[RelativePath, Document],
```
Comment (gabro) [inline language-server/src/main/scala/scala/meta/languageserver/SymbolIndexer.scala]: Why `java.util.Map`? Performance?

Comment (gabro) [reply language-server/src/main/scala/scala/meta/languageserver/SymbolIndexer.scala]: ah, I see, you're using `ConcurrentHashMap` in the implementation.

File: language-server/src/main/scala/scala/meta/languageserver/SymbolIndexer.scala [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+    ]
+) {
+
+  def documentSymbols(
+      path: RelativePath
+  ): Seq[(Position.Range, Denotation)] =
+    for {
+      document <- Option(documents.get(path)).toList
+      _ <- isFreshSemanticdb(path, document).toList
+      ResolvedName(pos: Position.Range, symbol: Symbol.Global, true) <- document.names
```
Comment (gabro) [inline language-server/src/main/scala/scala/meta/languageserver/SymbolIndexer.scala]: does this pattern act as a filter?

**Comment by gabro** [comment]
There are some errore that I'm struggling to minimize: essentially I can consistently break the "go to definition" feature with enough compile/clean/edit.

Also (very minor) I think we should skip `package` symbols in the "go to definition". I don't think there's an obvious place where to jump to and currently it does something very arbitrary.

File: language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala [hunk 7]
```diff
@@ -66,6 +76,7 @@ class ScalametaLanguageServer(
     logger.info(s"Initialized with $cwd, $pid, $rootPath, $capabilities")
     toCancel += scalafix.linter.subscribe()
     toCancel += symbol.indexer.subscribe()
+    loadAllSemanticdbsInWorkspace()
```
Comment (olafurpg) [inline language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala]: Should we consider offloading this to a separate thread?

File: language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala [hunk 8]
```diff
@@ -58,6 +58,16 @@ class ScalametaLanguageServer(
...
   private val toCancel = ListBuffer.empty[Cancelable]
 
+  private def loadAllSemanticdbsInWorkspace() = {
+    Files
+      .walk(workspacePath.toNIO)
+      .filter(
+        path =>
+          Files.isRegularFile(path) && path.toString.endsWith(".semanticdb")
+      )
+      .forEach(path => semanticdbSubscriber.onNext(AbsolutePath(path)))
```
Comment (olafurpg) [inline language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala]: We might want to do this on parallel, could be a lot of semanticdb files.

**Comment by olafurpg** [comment]
I think this PR is ready for merge. We definitely need to come up with a nice test fixture to reproduce bugs.

I did experience several issues when opening/closing vscode (to reload plugin) that the indexer doesn't always pick up everything. Indexing everything on startup helps a lot, however. I hope we can find a way to make it more robust.

File: language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala (hunk 8 above)
Comment (gabro) [reply language-server/src/main/scala/scala/meta/languageserver/ScalametaLanguageServer.scala]: right, that makes sense -> #11

**Comment by gabro** [comment]
I agree this is mature enough, let's track down bugs in future issues/PRs.

Merging 🎉

**Pull request merged**